apps/angular/crud/src/app/state/effects/todo.effects.ts

```diff
@@ -0,0 +1,106 @@
+import { Injectable, inject } from '@angular/core';
+import { Actions, createEffect, ofType } from '@ngrx/effects';
+import { TodoService } from '../../service/todo.service';
+import { map, exhaustMap, catchError, concatMap } from 'rxjs/operators';
+import { EMPTY, concat, of } from 'rxjs';
+import { TodoActionName } from '../actions/todo.actions';
+import { Todo } from '../../model/todo.interface';
+
+@Injectable()
+export class TodoEffects {
+  private actions$: Actions = inject(Actions);
+  private todoService: TodoService = inject(TodoService);
+
+  loadTodos$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(TodoActionName.callTodoList),
+      exhaustMap(() =>
+        this.todoService.callTodoList().pipe(
+          map((todoList) => ({
+            type: TodoActionName.loadTodoList,
+            todoList: todoList,
+          })),
+          catchError(() => EMPTY),
+        ),
+      ),
+    ),
+  );
+
+  updateTodo$ = createEffect(() =>
+    this.actions$.pipe(
+      ofType(TodoActionName.callUpdateTodo),
+      concatMap((action: { todo: Todo }) =>
+        concat(
+          of({
+            type: TodoActionName.todoStatus,
+            id: action.todo.id,
+            status: {
+              loading: true,
+            },
+          }),
+          this.todoService.updateTodo(action.todo).pipe(
+            map((todo) => ({
+              type: TodoActionName.updateTodoSuccess,
+              todo: todo,
+            })),
+            catchError(() =>
+              of({
+                type: TodoActionName.todoStatus,
+                id: action.todo.id,
+                status: {
+                  errorMsg: 'Error updating todo',
+                },
+              }),
+            ),
+          ),
+          of({
+            type: TodoActionName.todoStatus,
+            id: action.todo.id,
+            status: {
+              loading: false,
+            },
+          }),
+        ),
+      ),
+    ),
+  );
```

Comment:
I'm sorry but I don't understand why you are doing such a complicated effect, event flow ? Can you explain it to me. Where did you learn this ?

Reply:
Hello, I am learning ngrx with these challenges, I am going to review a better way to do it, thanks for the feedback

Reply:
> I'm sorry but I don't understand why you are doing such a complicated effect, event flow ? Can you explain it to me. Where did you learn this ?

Hello, I am learning ngrx with these challenges, I am going to review a better way to do it, thanks for the feedback

Reply:
Alright 💪 

An effect should return only one action and the reducer will update the state depending on your action, for instance loading or error.